WordPress/src/main/java/org/wordpress/android/viewmodel/posts/PostListItemDataSource.kt

```diff
@@ -0,0 +1,112 @@
+package org.wordpress.android.viewmodel.posts
+
+import org.wordpress.android.fluxc.Dispatcher
+import org.wordpress.android.fluxc.generated.PostActionBuilder
+import org.wordpress.android.fluxc.model.LocalOrRemoteId
+import org.wordpress.android.fluxc.model.LocalOrRemoteId.LocalId
+import org.wordpress.android.fluxc.model.LocalOrRemoteId.RemoteId
+import org.wordpress.android.fluxc.model.PostModel
+import org.wordpress.android.fluxc.model.SiteModel
+import org.wordpress.android.fluxc.model.list.PostListDescriptor
+import org.wordpress.android.fluxc.model.list.datasource.ListItemDataSourceInterface
+import org.wordpress.android.fluxc.store.PostStore
+import org.wordpress.android.fluxc.store.PostStore.FetchPostListPayload
+import org.wordpress.android.fluxc.store.PostStore.RemotePostPayload
+import org.wordpress.android.viewmodel.posts.PostListItemIdentifier.EndListIndicatorIdentifier
+import org.wordpress.android.viewmodel.posts.PostListItemIdentifier.LocalPostId
+import org.wordpress.android.viewmodel.posts.PostListItemIdentifier.RemotePostId
+import org.wordpress.android.viewmodel.posts.PostListItemType.EndListIndicatorItem
+import org.wordpress.android.viewmodel.posts.PostListItemType.LoadingItem
+import org.wordpress.android.viewmodel.posts.PostListItemType.PostListItemUiState
+
+sealed class PostListItemIdentifier {
+    data class LocalPostId(val id: LocalId) : PostListItemIdentifier()
+    data class RemotePostId(val id: RemoteId) : PostListItemIdentifier()
+    object EndListIndicatorIdentifier : PostListItemIdentifier()
+}
+
+class PostListItemDataSource(
+    private val dispatcher: Dispatcher,
+    private val postStore: PostStore,
+    private val performGetItemIdsToHide: ((PostListDescriptor) -> Pair<LocalPostId, RemotePostId>?),
+    private val transform: ((PostModel) -> PostListItemUiState)
+) : ListItemDataSourceInterface<PostListDescriptor, PostListItemIdentifier, PostListItemType> {
+    override fun fetchList(listDescriptor: PostListDescriptor, offset: Long) {
+        val fetchPostListPayload = FetchPostListPayload(listDescriptor, offset)
+        dispatcher.dispatch(PostActionBuilder.newFetchPostListAction(fetchPostListPayload))
+    }
+
+    override fun getItemIdentifiers(
+        listDescriptor: PostListDescriptor,
+        remoteItemIds: List<RemoteId>,
+        isListFullyFetched: Boolean
+    ): List<PostListItemIdentifier> {
+        val localItems = postStore.getLocalPostIdsForDescriptor(listDescriptor)
+                .map { LocalPostId(id = it) }
+        val remoteItems = remoteItemIds.map { RemotePostId(id = it) }
+        val actualItems = localItems + remoteItems
+
+        val endListItem = if (isListFullyFetched && actualItems.isNotEmpty()) {
+            listOf(EndListIndicatorIdentifier)
+        } else emptyList()
+        val itemsToHide = getItemIdsToHide(listDescriptor)
+        // We only want to show the end list indicator if the list is fully fetched and it's not empty
```

Comment:
Nitpick: Shouldn't this comment be moved 4 lines above?

Reply:
Fixed in 093558595838c2a8219b6991cb9e1bb4bbc48e62.